proposals/0082-relax-transaction-constraints.md

```diff
@@ -0,0 +1,173 @@
+relaxed as well.
+
+The relaxation of these constraints is only relaxed for transactions being
+included in a block. During block validation, if any of these constraints
+are broken, the entire block is marked invalid; with this proposal, the
+block is not marked invalid, but the transaction will not be executed.
+These constraints must still be satisfied in order for the transaction to be
+executed, and have an effect on state. However, there are some new considerations
+which must be taken into account, specifically as it relates to fees and block
+limits.
+
+### Fee-Paying
+
+Currently, iff a transaction's fee-payer does not have enough funds to pay the
+fee, the transaction cannot be included in a block. With this proposal, it is
+possible for such a transaction to be included in the block, and there are
+three different edge-cases to consider:
+
+1. The fee-payer account does not exist (0 lamports)
+2. The fee-payer account does not have enough funds for the entire fee
+3. The fee-payer account has enough funds for the fee, but would no longer be
+rent-exempt
+
+In case 1, the transaction should simply be ignored for execution, and have no
+effect on state.
+In case 2, the fee-paying account should be drained of all funds, but have no
+other effect on state.
+In case 3, the fee-paying account should be drained of all funds, with fees
+being paid to the block-producer, and the remainder of lamports being dropped.
+
+### Block-Limits
+
+Pending the activation of `2ry7ygxiYURULZCrypHhveanvP5tzZ4toRwVp89oCNSj`,
+validators must validate a block is within block-limits.
+With this proposal, some transactions may not be executable, but will still
+count towards block-limits.
+If these transactions did not count towards block-limits, the validation of
+block-limits would require the validator to check whether or not a transaction
+is executable, which negates the benefits of this proposal.
+
+Additionally, f these transactions did not count towards block-limits, a
+malicious leader could produce a block with non-executable transactions and
+overload the network.
+
+## Impact
+
+- Transactions that would previously be dropped with an error, can now be
+  included, and even charged fees.
+  - Users must be more careful when constructing transactions to ensure they
+    are executable if they don't want to waste fees
+- The validity of a block is no longer dependent on intra-block account-state
+updates. This is because the only account-state required for transaction
+validation is the ALT resolution, which is resolved at the beginning of a slot.
+  - Block-production could be done asynchronously
+  - Block-validation could be done without execution, but still relies on the
+  execution of previous blocks.
+
+## Security Considerations
+
+- Removing fee-paying requirement could allow validator clients to produce
+  blocks that do not pay fees. If not checked by block-producer, this could
+  allow malicious users to abuse such producers/leaders.
+
```

Comment:
also, malicious leaders could create full of garbage blocks, which only contains bad transactions at no cost, burdening all the validations onto other validators, esp, heavy account state lookup.

Reply:
agreed with @ryoqun, this attack vector is the main thing that needs to be solved. Maybe this specific problem can be discussed in a separate SIMD since relaxing other transaction constraints is more straightforward.

Reply:
made a comment in this chain: https://github.com/solana-foundation/solana-improvement-documents/pull/82#discussion_r1393361600

I don't think there's a DOS that's creating more work than would be possible within current protocol.
I can currently create a block thats full of **valid** heavy transactions, and then tag in some invalidating transaction at the last tick. All the validators wasted their time executing all those txs, and there's no reward for them since the block is thrown out entirely.

This would change the vector, since they could have **all** txs that do not pay fees. But, that only makes the execution less intense, since they still have a cost that counts towards block limits.